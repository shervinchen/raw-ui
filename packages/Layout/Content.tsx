import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { ContentProps } from './Layout.types';

const Content: FC<PropsWithChildren<ContentProps>> = ({
  className = '',
  children,
  ...restProps
}) => {
  const classes = classNames('raw-layout-content', className);

  return (
    <main className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-layout-content {
          flex: auto;
        }
      `}</style>
    </main>
  );
};

export default Content;
