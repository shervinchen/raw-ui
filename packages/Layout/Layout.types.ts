import { ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

interface BaseLayoutProps {
  className?: string;
}

interface BaseHeaderProps {
  className?: string;
}

interface BaseContentProps {
  className?: string;
}

interface BaseFooterProps {
  className?: string;
}

interface BaseSidebarProps {
  className?: string;
}

export type LayoutProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseLayoutProps
>;
export type HeaderProps = Merge<
  ComponentPropsWithoutRef<'header'>,
  BaseHeaderProps
>;
export type ContentProps = Merge<
  ComponentPropsWithoutRef<'main'>,
  BaseContentProps
>;
export type FooterProps = Merge<
  ComponentPropsWithoutRef<'footer'>,
  BaseFooterProps
>;
export type SidebarProps = Merge<
  ComponentPropsWithoutRef<'aside'>,
  BaseSidebarProps
>;
