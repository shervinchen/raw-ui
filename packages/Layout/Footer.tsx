import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { FooterProps } from './Layout.types';

const Footer: FC<PropsWithChildren<FooterProps>> = ({
  className = '',
  children,
  ...restProps
}) => {
  const classes = classNames('raw-layout-footer', className);

  return (
    <footer className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-layout-footer {
          flex: 0 0 auto;
        }
      `}</style>
    </footer>
  );
};

export default Footer;
