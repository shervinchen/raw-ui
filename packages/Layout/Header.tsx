import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { HeaderProps } from './Layout.types';

const Header: FC<PropsWithChildren<HeaderProps>> = ({
  className = '',
  children,
  ...restProps
}) => {
  const classes = classNames('raw-layout-header', className);

  return (
    <header className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-layout-header {
          flex: 0 0 auto;
        }
      `}</style>
    </header>
  );
};

export default Header;
