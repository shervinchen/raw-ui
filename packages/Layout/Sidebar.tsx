import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { SidebarProps } from './Layout.types';

const Sidebar: FC<PropsWithChildren<SidebarProps>> = ({
  className = '',
  children,
  ...restProps
}) => {
  const classes = classNames('raw-layout-sidebar', className);

  return (
    <aside className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-layout-sidebar {
          position: relative;
          transition: all 0.2s;
        }
      `}</style>
    </aside>
  );
};

export default Sidebar;
