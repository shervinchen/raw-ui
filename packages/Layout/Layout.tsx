import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { LayoutProps } from './Layout.types';
import { getValidChildren } from '../utils/common';
import Sidebar from './Sidebar';

const Layout: FC<PropsWithChildren<LayoutProps>> = ({
  className = '',
  children,
  ...restProps
}) => {
  const hasSidebar = getValidChildren(children).some(
    (child) => child.type === Sidebar
  );
  const classes = classNames(
    'raw-layout',
    hasSidebar && 'raw-layout-has-sidebar',
    className
  );

  return (
    <div data-testid="layout" className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-layout {
          display: flex;
          flex: auto;
          flex-direction: column;
        }
        .raw-layout-has-sidebar {
          flex-direction: row;
        }
      `}</style>
    </div>
  );
};

export default Layout;
