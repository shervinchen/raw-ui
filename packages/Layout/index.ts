import Layout from './Layout';
import Header from './Header';
import Content from './Content';
import Sidebar from './Sidebar';
import Footer from './Footer';

export type LayoutComponentType = typeof Layout & {
  Header: typeof Header;
  Content: typeof Content;
  Sidebar: typeof Sidebar;
  Footer: typeof Footer;
};

(Layout as LayoutComponentType).Header = Header;
(Layout as LayoutComponentType).Content = Content;
(Layout as LayoutComponentType).Sidebar = Sidebar;
(Layout as LayoutComponentType).Footer = Footer;

export default Layout as LayoutComponentType;
