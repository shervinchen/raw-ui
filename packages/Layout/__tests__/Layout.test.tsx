import React from 'react';
import { render, screen } from '@testing-library/react';
import Layout from '..';

describe('Layout', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Layout>
        <Layout.Header>Header</Layout.Header>
        <Layout>
          <Layout.Sidebar>Sidebar</Layout.Sidebar>
          <Layout.Content>Content</Layout.Content>
        </Layout>
        <Layout.Footer>Footer</Layout.Footer>
      </Layout>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(
      <Layout className="custom-layout">
        <Layout.Header className="custom-layout-header">Header</Layout.Header>
        <Layout>
          <Layout.Sidebar className="custom-layout-sidebar">
            Sidebar
          </Layout.Sidebar>
          <Layout.Content className="custom-layout-content">
            Content
          </Layout.Content>
        </Layout>
        <Layout.Footer className="custom-layout-footer">Footer</Layout.Footer>
      </Layout>
    );
    expect(screen.getAllByTestId('layout')[0]).toHaveClass('custom-layout');
    expect(screen.getByTestId('layoutHeader')).toHaveClass(
      'custom-layout-header'
    );
    expect(screen.getByTestId('layoutSidebar')).toHaveClass(
      'custom-layout-sidebar'
    );
    expect(screen.getByTestId('layoutContent')).toHaveClass(
      'custom-layout-content'
    );
    expect(screen.getByTestId('layoutFooter')).toHaveClass(
      'custom-layout-footer'
    );
  });

  test('should detect the sidebar as children', () => {
    render(
      <Layout>
        <Layout.Header>Header</Layout.Header>
        <Layout>
          <Layout.Sidebar>Sidebar</Layout.Sidebar>
          <Layout.Content>Content</Layout.Content>
        </Layout>
        <Layout.Footer>Footer</Layout.Footer>
      </Layout>
    );
    expect(screen.getAllByTestId('layout')[1]).toHaveClass(
      'raw-layout-has-sidebar'
    );
  });
});
