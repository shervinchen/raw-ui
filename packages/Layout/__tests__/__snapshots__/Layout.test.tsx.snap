// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-813210850 raw-layout"
    data-testid="layout"
  >
    <header
      class="jsx-4037815597 raw-layout-header"
      data-testid="layoutHeader"
    >
      Header
    </header>
    <div
      class="jsx-813210850 raw-layout raw-layout-has-sidebar"
      data-testid="layout"
    >
      <aside
        class="jsx-2255274172 raw-layout-sidebar"
        data-testid="layoutSidebar"
      >
        Sidebar
      </aside>
      <main
        class="jsx-1245817261 raw-layout-content"
        data-testid="layoutContent"
      >
        Content
      </main>
    </div>
    <footer
      class="jsx-2757715662 raw-layout-footer"
      data-testid="layoutFooter"
    >
      Footer
    </footer>
  </div>
</DocumentFragment>
`;
